File: LeetCode_Java/twopoint/PalindromicSubstrings.java
package twopoint;

public class PalindromicSubstrings {
    public int countSubstrings(String s) {

        int result = 0;

        for (int i = 0; i < s.length(); i++) {
            result += countPalindromic(s, i, i);
            result += countPalindromic(s, i, i + 1);
        }

        return result;
    }

    private int countPalindromic(String s, int left, int right) {
        int count = 0;
        while (left >= 0 && right < s.length()) {
            if (s.charAt(left) != s.charAt(right)) {
                break;
            }
            left--;
            right++;
            count++;
        }

        return count;
    }

    public static void main(String[] args) {
        PalindromicSubstrings solution = new PalindromicSubstrings();
        System.out.println(solution.countSubstrings("abc")); // Output: 3
        System.out.println(solution.countSubstrings("aaa")); // Output: 6
        System.out.println(solution.countSubstrings("abacdfgdcaba")); // Output: 14
    }
}

File: LeetCode_Python/twopoint/PalindromicSubstrings.py
class PalindromicSubstrings:
    def countSubstrings(self, s: str) -> int:
        n = len(s)

        def count(left: int, right: int) -> int:
            count = 0

            while left >= 0 and right < n:
                if s[left] != s[right]:
                    break
                count += 1
                left -= 1
                right += 1

            return count

        res = 0

        for i in range(n):
            res += count(i, i)
            res += count(i, i + 1)

        return res


if __name__ == "__main__":
    solution = PalindromicSubstrings()

    print("Expected Output : 3")
    print("Actual Output : ", solution.countSubstrings("abc"))

    print("Expected Output : 6")
    print("Actual Output : ", solution.countSubstrings("aaa"))

    print("Expected Output : 6")
    print("Actual Output : ", solution.countSubstrings("abba"))
